pkg/awsutils/awsutils.go

```diff
@@ -442,20 +451,61 @@ func (cache *EC2InstanceMetadataCache) getENIMetadata(macStr string) (ENIMetadat
 	}
 	log.Debugf("Found ENI: %s, MAC %s, device %d", eni, eniMAC, deviceNum)
 
-	localIPv4s, cidr, err := cache.getIPsAndCIDR(eniMAC)
+	imdsIPv4s, cidr, err := cache.getIPsAndCIDR(eniMAC)
 	if err != nil {
 		return ENIMetadata{}, errors.Wrapf(err, "get ENI metadata: failed to retrieve IPs and CIDR for ENI: %s", eniMAC)
 	}
-	_, tags, _, err := cache.DescribeENI(eni)
+	privateIPv4s, tags, _, err := cache.DescribeENI(eni)
 	if err != nil {
 		return ENIMetadata{}, errors.Wrapf(err, "get ENI metadata: failed to describe ENI: %s, %v", eniMAC, err)
 	}
+	// getIPsAndCIDR() queries IMDS for IPv4 addresses attached to the ENI.
+	// DescribeENI() calls the DescribeNetworkInterfaces AWS API call, which
+	// technically should be the source of truth and contain the freshest
+	// information. Let's just do a quick scan here and output some diagnostic
+	// messages if we find stale info in the IMDS result.
+	missingIMDS := []string{}
+	missingDNI := []string{}
+	for _, privateIPv4 := range privateIPv4s {
+		found := false
+		strPrivateIPv4 := aws.StringValue(privateIPv4.PrivateIpAddress)
+		for _, imdsIPv4 := range imdsIPv4s {
+			if imdsIPv4 == strPrivateIPv4 {
+				found = true
+				break
+			}
+		}
+		if !found {
+			missingIMDS = append(missingIMDS, strPrivateIPv4)
+		}
+	}
+	for _, imdsIPv4 := range imdsIPv4s {
+		found := false
+		for _, privateIPv4 := range privateIPv4s {
+			strPrivateIPv4 := aws.StringValue(privateIPv4.PrivateIpAddress)
+			if imdsIPv4 == strPrivateIPv4 {
+				found = true
+				break
+			}
+		}
+		if !found {
+			missingDNI = append(missingDNI, imdsIPv4)
+		}
+	}
```

Comment:
Working with sets in go is somewhat painful... 😉

Reply:
indeed. there is a `sets` utility library in upstream k/k that makes working with this a bit easier, but wasn't worth pulling in the dep IMHO.

Reply:
@jaypipes Nick added "github.com/deckarep/golang-set" earlier.. :)